flink-state-backends/flink-statebackend-forst/src/main/java/org/apache/flink/state/forst/fs/filemapping/FileMappingManager.java

```diff
@@ -0,0 +1,250 @@
+        for (Map.Entry<String, MappingEntry> entry : mappingTable.entrySet()) {
+            if (isParentDir(entry.getKey(), path)) {
+                linkedPaths.add(entry.getKey());
+            }
+        }
+        return linkedPaths;
+    }
+
+    /**
+     * 1. If src can match any key, we only `mark rename`, no physical file would be renamed. 2. If
+     * src is a directory, all files under src will be renamed, including linked files and local
+     * files, the directory also would be renamed in file system physically.
+     *
+     * @param src the source path
+     * @param dst the destination path
+     * @return always return true except for IOException
+     */
+    public boolean renameFile(String src, String dst) throws IOException {
+        MappingEntry srcEntry = mappingTable.get(src);
+        if (srcEntry != null) { // rename file
+            if (mappingTable.containsKey(dst)) {
+                MappingEntry dstEntry = mappingTable.remove(dst);
+                dstEntry.release();
+            }
+            mappingTable.remove(src);
+            mappingTable.put(dst, srcEntry);
+        } else { // rename directory = link to dst dir + delete src dir
+
+            // step 1: link all files under src to dst
+            List<String> toRename = listByPrefix(src);
+            for (String key : toRename) {
+                MappingEntry sourceEntry = mappingTable.get(key);
+                sourceEntry.retain();
+                String renamedDst = key.replace(src, dst);
+                LOG.trace("rename: {} -> {}", key, renamedDst);
+                mappingTable.put(renamedDst, sourceEntry);
+            }
+
+            Path dstPath = new Path(dst);
+            if (!fileSystem.exists(dstPath)) {
+                fileSystem.mkdirs(dstPath);
+            }
+            // step 2: delete src dir
+            deleteFile(new Path(src), true);
+        }
+        return true;
+    }
+
+    /**
+     * Delete a file or directory from mapping table and file system, the directory deletion may be
+     * deferred.
+     *
+     * @param file to be deleted
+     * @param recursive whether to delete recursively
+     * @return true if the file or directory is deleted successfully, false otherwise.
+     * @throws IOException if an error occurs during deletion
+     */
+    public boolean deleteFile(Path file, boolean recursive) throws IOException {
+        String fileStr = file.toString();
+        MappingEntry entry = mappingTable.getOrDefault(fileStr, null);
+        LOG.trace("delete: {}, source:{}", file, entry == null ? "null" : entry.sourcePath);
+        // case 1: delete file
+        if (entry != null) {
+            mappingTable.remove(fileStr);
+            entry.release();
+            return true;
+        }
+
+        // case 2: delete directory
+        MappingEntry parentEntry = new MappingEntry(0, fileSystem, fileStr, false, recursive);
+
+        boolean matched = false;
```

Comment:
I suggest removing this `matched`, and delete the directory if `parentEntry.getReferenceCount()` is true in line 211 (must before step 2.2).